### DISCUSSION TITLE
cargo crev discussion

### DISCUSSION THREAD

**Discussion by apoelstra** [rust-bitcoin/rust-bitcoin#1396] [General]
Creating this to share notes and thoughts about `cargo crev` . I installed it the other day and poked around a bit. It's very featureful and easy to navigate (in that you can stick `help` onto the end of all the commands to discover more functionality). I used [this walkthrough](https://wiki.alopex.li/ActuallyUsingCrev) which also has a bunch of commentary, though it seems like the specific commands are a bit out of date, since I had to tweak them somewhat.

I have a few rust-bitcoin specific thoughts:

* Should I/we be reviewing our own crates? This seems a little misleading because if there are security problems here, presumably they correspond to my own blind spots, otherwise I'd fix them. OTOH if I were to review my own releases that'd give a signature attesting that they're really from me which is in the cargo/crev WoT. We have signed tags right now but it's a big PITA to match git tags to cargo releases. I understand there's some magic file in cargo checkouts with a git commit ID, but sometimes this isn't present and anyway even with the file you've gotta do an out-of-band check that the commit ID actually matches the checkout.
* What `cargo crev` tooling should we support in our CI or in our `contrib/` directory? I think we should have the full output of `cargo crev` committed somewhere, and have a script which does `cargo update -p --precise` a bunch of times to bring all deps to the most-recently-reviewed versions. Or something like that. This output should be generated from a `crev` configuration/WoT that we also commit, so there's an explicit list of people who are trusted.
* Relatedly we should commit the output of `cargo-supply-chain` which lists everybody who is able to push to crates.io for anything in our dependency tree. I could have a cronjob or something that alerted me if the output ever changed, and we could PR to change it and have a discussion (if necessary).

It seems a bit optimistic to be able to review every dependency ... in particular `libc` is a **massive** crate and has frequent minor releases with multi-1000-line diffs. These mostly just rearrange data structures, and the authors are widely trusted Rust OGs, but that doesn't translate into a trustworthy crev review. (Interesting, if *they* would review their own crates, that would be super helpful since I'd be willing to put joshtriplett or huonw into the WoT, say.)

As a followup discussion, if we did all this we could have very different discussions about adding dependencies. Rather than being basically "hard no unless there's a very good reason" we could have a more nuanced policy where we'd be willing to add well-maintained dependencies belonging to people we already trusted, or crates which had regular crev reviews from people we trust, or which were small enough that *we* could sustainably review every release, etc.

cc @dpc in case he's not subscribed to this repo
cc @TheBlueMatt who cares a lot about supply chain integrity

**Comment by apoelstra** [comment]
BTW the current supply-chain output is as follows. I think this is really good: of the 17 individuals listed, I count 6 Rust maintainers, 4 I am or know personally, and 5 are people I don't know/only know from blog posts, but only control dev-dependencies. The unknown names mostly control widely-used crates (e.g. `libc`). The only thing that concerns me at all is `ppv-lite86`, a dependency of `rand-chacha` where I don't know the author and worry that this crate is too niche to have many eyeballs on it.

If I compile without `rand` and without `wasi` (which I think is only pulled in if I'm doing wasm builds anyway) then the list is 100% Rust maintainers and friends of mine :)

```
17:33:50 -bash@camus ~/code/rust-bitcoin/rust-bitcoin/master 70eb92cc8$ cargo supply-chain publishers

The following crates will be ignored because they come from a local directory:
 - bitcoin
 - bitcoin-internals

Using cached data. Cache age: 16 hours

The following individuals can publish updates for your dependencies:

 1. alexcrichton via crates: cfg-if, libc, rand, rand_core, wasi
 2. dtolnay via crates: itoa, ryu, serde, serde_json, serde_test
 3. apoelstra via crates: bech32, bitcoin_hashes, secp256k1, secp256k1-sys
 4. dhardy via crates: getrandom, rand, rand_chacha, rand_core
 5. rust-lang-owner via crates: cc, cfg-if, libc
 6. TheBlueMatt via crates: bech32, bitcoin_hashes
 7. JohnTitor via crates: libc
 8. TyOverby via crates: bincode
 9. VictorKoenders via crates: bincode
 10. ZoeyR via crates: bincode
 11. clarkmoody via crates: bech32
 12. gnzlbg via crates: libc
 13. huonw via crates: libc
 14. jonasnick via crates: secp256k1
 15. joshtriplett via crates: libc
 16. kazcw via crates: ppv-lite86
 17. sunfishcode via crates: wasi

Note: there may be outstanding publisher invitations. crates.io provides no way to list them.
See https://github.com/rust-lang/crates.io/issues/2868 for more info.

All members of the following teams can publish updates for your dependencies:

 1. "github:rust-random:maintainers" (https://github.com/rust-random) via crates: getrandom, rand, rand_chacha, rand_core
 2. "github:rust-lang:libs" (https://github.com/rust-lang) via crates: cc, cfg-if, libc
 3. "github:serde-rs:owners" (https://github.com/serde-rs) via crates: serde, serde_json, serde_test
 4. "github:rust-lang-nursery:libs" (https://github.com/rust-lang-nursery) via crates: rand
 5. "github:rust-lang:libc" (https://github.com/rust-lang) via crates: libc

Github teams are black boxes. It's impossible to get the member list without explicit permission.
```

**Reply by dpc** [reply]
BTW. Probably the best corresponding output from `cargo-crev` would be:

```
> cargo crev verify --show-all
status reviews issues owner      downloads    loc lpidx geiger flgs crate                      version                       latest_t    
none     0   2  0   0  2  3  2720K  28502K    202   159      0 CB__ wasm-bindgen-shared        0.2.83                        ↓0.2.76     
none     0   1  0   0  1  1   533K  23303K    134   144      6 ____ scoped-tls                 1.0.1                         ↓1.0.0      
none     0   9  0   0  1  1  6389K 120918K    262   275      7 ____ itoa                       1.0.4                         ↓0.4.8      
pass     5  12  0   0  2  3 88515K 154633K    118   411      0 ____ cfg-if                     1.0.0                         =           
pass     1   2  0   0  0  3   566K   3793K    758    35      0 ____ bech32                     0.9.1                         =           
none     0   8  0   0  1  1 12653K  96804K   2107    84    734 ____ ryu                        1.0.11                        ↓1.0.5      
none     0   2  0   0  0  2  4549K  91349K   1035   129    135 ____ once_cell                  1.16.0                        ↓1.8.0      
none     1   1  0   0  0  2   279K   2003K   3993    10     56 ____ bitcoin_hashes             0.11.0                        =           
none     0   0  0   0  1  2  7910K  46947K   1526    75    698 ____ wasi                       0.11.0+wasi-snapshot-preview1             
none     0   1  0   0  0  1  2435K  78866K   3522    57    739 ____ ppv-lite86                 0.2.17                        ↓0.2.6      
none     0   0  0   0  1  1  7502K  31293K    754   146      4 ____ unicode-ident              1.0.5                                     
none     0  12  0   0  2  2  1587K  29160K   3559    39   1201 ____ bumpalo                    3.11.1                        ↓3.8.0      
none     0   2  0   0  5  7 23613K 118587K   3094    70     20 CB__ log                        0.4.17                        ↓0.4.8      
none     0  19  0   0  1  2    86K 103230K   4038    60    208 ____ cc                         1.0.78                        ↓1.0.71     
local    0   0  0   0  ?  ?        ?         ?    501     0     10 CB__ bitcoin-internals          0.1.0*                                    
none     0  20  0   0  1  1  6983K 152263K   3459    78     18 CB__ proc-macro2                1.0.47                        ↓1.0.30     
none     0   4  0   0  1  2   439K 132069K  10283    43      5 CB__ serde                      1.0.150                       ↓1.0.130    
pass     1   1  0   0  0  1   245K   4449K   1223    27    776 CB__ secp256k1-sys              0.6.1                         =           
none     0   0  0   0  1  2     5K   7052K   2014    17      0 CB__ serde_test                 1.0.150                                   
none     0   1  0   0  0  3 15527K  25030K   2904    36     22 ____ bincode                    1.3.3                         ↓1.1.4      
none     0   1  0   0  1  2  2031K 110520K  12098    37      7 CB__ serde_json                 1.0.89                        ↓1.0.66     
none     0  14  0   0  1  1 16446K 154778K   1370   122      0 CB__ quote                      1.0.21                        ↓1.0.10     
none     0  72  0   0  4  8  1295K 156489K  88661    15    477 CB__ libc                       0.2.138                       ↓0.2.137    
none     0   8  0   0  2  2  4024K 116906K    969   135    183 ____ getrandom                  0.2.8                         ↓0.2.3      
none     0  10  0   0  3  3  6489K 158175K    829   154      2 ____ rand_core                  0.6.4                         ↓0.6.3      
none     0   1  0   0  2  3   301K   3961K     69    96      0 ____ wasm-bindgen-test-macro    0.3.33                        ↓0.3.26     
none     0   8  0   0  1  1  1496K 168643K  37584    24     74 CB__ syn                        1.0.105                       ↓1.0.80     
pass     1   6  0   0  2  2 43663K 119216K    711   147      0 ____ rand_chacha                0.3.1                         =           
none     0   1  0   0  2  3  2716K  28508K   2151    49      0 ____ wasm-bindgen-backend       0.2.83                        ↓0.2.76     
none     0  12  0   0  4  4 25019K 166321K   5273    59     32 ____ rand                       0.8.5                         ↓0.8.4      
none     0   1  0   0  1  2   402K 113211K   7216    48      0 CB__ serde_derive               1.0.150                       ↓1.0.130    
none     0   1  0   0  2  3  2716K  28457K   1544    58      0 ____ wasm-bindgen-macro-support 0.2.83                        ↓0.2.76     
pass     1   3  0   0  0  2    28K   5640K   4562    15    446 ____ secp256k1                  0.24.2                        ↑0.25.0     
none     0   1  0   0  2  3  2716K  28499K    299   131      1 ____ wasm-bindgen-macro         0.2.83                        ↓0.2.76     
local    0   3  0   0  0  2     0K   1325K  23415     3     24 CB__ bitcoin                    0.29.0*                       ↑0.29.2     
none     0   1  0   0  2  3  2721K  28615K   2898    42    541 CB__ wasm-bindgen               0.2.83                        ↓0.2.76     
pass     1   3  0   0  2  2  3829K   6831K     44   151      0 ____ console_error_panic_hook   0.1.7                         =           
none     0  14  0   0  2  3  2552K  27017K   2863    42     10 ____ js-sys                     0.3.60                        ↓0.3.55     
none     0   1  0   0  2  3  1995K  24318K 163333     5      0 ____ web-sys                    0.3.60                        ↓0.3.53     
none     0   1  0   0  2  3  1525K  19942K    494    81     70 ____ wasm-bindgen-futures       0.4.33                        ↓0.4.26     
none     0   1  0   0  2  3   308K   4028K    441    38      4 ____ wasm-bindgen-test          0.3.33                        ↓0.3.26     
local    1   1  0   0  0  2   279K   2003K   4160    10     59 ____ bitcoin_hashes             0.11.0*                       =           
```

It includes some data from crates.io as well. (`--help` shows a description of what each colum means).

Also - there is some support for whitelisting known and reputable users from crates.io (`cargo crev repo edit known`) but it's a half-baked feature that was developed long time ago and it's usefulness not really decided on. Probably to make it more useful, the file should be "current project"-scoped, and stored somewhere in the source code itself, and not local to the user.

**Comment by dpc** [comment]
> Should I/we be reviewing our own crates? This seems a little misleading because if there are security problems here, presumably they correspond to my own blind spots, otherwise I'd fix them. 

Historically the answer was always yes. Primary reason is that we can't really prevent it, so might as well make it a standard behavior that downstream users should be aware of. But there are also other benefits:

* Each review contains a cryptographic checksum of the package content, so downstream users get some minimum level of protection just because of that.
* If someone already trusts you (possibly transitively), the information that you reviewed it carries similar meaning to you being the author. They probably trust that one way or another. At least right now the primary goal is more about coordinating people with pre-existing trust relationships into prioritizing reviewing things authored by people they don't know&trust.


> What cargo crev tooling should we support in our CI or in our contrib/ directory

For CI use the idea was to have:

`cargo crev verify --for-id <ID>`

where ID is a public ID of someone inside the project (root of trust) and require that every dependency is approved before landing. Same command (with same ID) can be put somewhere in `./contrib`

> so there's an explicit list of people who are trusted

That ID above should just publish trust proofs for all trusted users. Such yaml files can be even embedded in the repository itself.

> It seems a bit optimistic to be able to review every dependency ... in particular libc is a massive crate and has frequent minor releases with multi-1000-line diffs.

That's what `understanding` and `thoroughness` levels are in review. It's OK to rubber-stump a dependency with `thoroughness: none;` and if someone downstream your WoT does not want to take into account such reviews, they have command line options for all commands (like `cargo crev verify --thoroughness low`) to filter them out.

**Reply by apoelstra** [reply]
Thanks, these are super useful clarifications!

**Reply by Kixunil** [reply]
I definitely agree with the answer to "Should I/we be reviewing our own crates?". I originally wanted to create a super-secure Qubes setup for my keys but now that you started this I will probably go with something less paranoid but still reasonable just to get it done sooner.

> we could have a more nuanced policy

This sounds exciting AF!

**Reply by dpc** [reply]
Indeed I would generally advise to start small and aim low. After you reach certain level, you can keep tightening the requirements. Crev was designed to support this approach.

**Comment by tcharding** [comment]
I've published reviews for all our (current) crates depended on by `rust-bitcoin`. I did `byteorder` as well since its only 800 LOC.

**Reply by dpc** [reply]
It seems like you two IDs:

```
> cargo crev repo f url https://github.com/tcharding/crev-proofs
Fetching https://github.com/tcharding/crev-proofs... 
https://github.com/tcharding/crev-proofs                     new: 3 trust, 11 package reviews
Found proofs from:
      13 Df2R3MuKZTrPtFwmRvCfArOTY2E3IlKBJEEXqk_YMvY (verified owner)
       1 cplyiUAQysMfRsw1WCYPvhHWhr0fxH01OkuojK0SRyY (verified owner)
```

Is `Df2R3MuKZTrPtFwmRvCfArOTY2E3IlKBJEEXqk_YMvY` what I should assign trust for, or do you plan to use both?

**Reply by dpc** [reply]
```
> cargo crev repo git show
commit a60f5105d928bd564925e9f39c8dbe1c2b1bc037 (HEAD -> master, origin/master, origin/HEAD)
Author: Dawid Ciężarkiewicz <dpc@dpc.pw>
Date:   Thu Dec 15 20:56:09 2022 -0800

    Add trust for Df2R3MuKZTrPtFwmRvCfArOTY2E3IlKBJEEXqk_YMvY

diff --git a/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev b/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev
new file mode 100644
index 0000000..51af17b
--- /dev/null
+++ b/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev
@@ -0,0 +1,17 @@
+----- BEGIN CREV PROOF -----
+kind: trust
+version: -1
+date: 2022-12-15T20:56:09.422409249-08:00
+from:
+  id-type: crev
+  id: FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE
+  url: https://github.com/dpc/crev-proofs
+ids:
+- id-type: crev
+  id: Df2R3MuKZTrPtFwmRvCfArOTY2E3IlKBJEEXqk_YMvY
+  url: https://github.com/tcharding/crev-proofs
+trust: medium
+----- SIGN CREV PROOF -----
+l-qqI4xeCZo_WVdKBU4U8ZbTVdTr5FAs91inEjmhZUmAE3x7b7JZ2KbV8TxMOCEKJkkWuCBL33s8DALNwQFYCQ
+----- END CREV PROOF -----
+
```

It's official now. :D

**Reply by dpc** [reply]
Oh, and I don't think I ever made one for @apoelstra 

```
> cargo crev repo git show
commit 8ecc82d567dcb32fc93164a9a157c48eae86680b (HEAD -> master)
Author: Dawid Ciężarkiewicz <dpc@dpc.pw>
Date:   Thu Dec 15 20:58:02 2022 -0800

    Add trust for AOdw0m1s0DwuXFi3B2Juka2Wnx1LjJPD5qo3IfYTtlA

diff --git a/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev b/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev
index 51af17b..db4a5e3 100644
--- a/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev
+++ b/FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE/trust/2022-12-FJFJUQ.proof.crev
@@ -15,3 +15,20 @@ trust: medium
 l-qqI4xeCZo_WVdKBU4U8ZbTVdTr5FAs91inEjmhZUmAE3x7b7JZ2KbV8TxMOCEKJkkWuCBL33s8DALNwQFYCQ
 ----- END CREV PROOF -----
 
+----- BEGIN CREV PROOF -----
+kind: trust
+version: -1
+date: 2022-12-15T20:58:02.044758860-08:00
+from:
+  id-type: crev
+  id: FYlr8YoYGVvDwHQxqEIs89reKKDy-oWisoO0qXXEfHE
+  url: https://github.com/dpc/crev-proofs
+ids:
+- id-type: crev
+  id: AOdw0m1s0DwuXFi3B2Juka2Wnx1LjJPD5qo3IfYTtlA
+  url: https://github.com/apoelstra/crev-proofs
+trust: medium
+----- SIGN CREV PROOF -----
+ZOk23-ueALs3df6nrRAgAQACPmEn-9-r4oczC_TK2gnM7K4aOlQlUbH3GL2Txfs37olw0euc5z5Vz6ZVzF9UDA
+----- END CREV PROOF -----
```

**Reply by apoelstra** [reply]
I think I only published one (test) review so far. My goal was to go over everything over Christmas break :)

**Comment by tcharding** [comment]
> Is Df2R3MuKZTrPtFwmRvCfArOTY2E3IlKBJEEXqk_YMvY what I should assign trust for, or do you plan to use both?

I messed up and created the second one, I made and effort to clean up - I guess not well enough. I do not plan to use the other one, all published reviews, I think, are using the "main" one. This is starting to sound very much like my pgp key :(

**Reply by dpc** [reply]
@tcharding You can use `cargo crev repo git ...` like a normal `git`. So `cargo crev repo git ls` + `cargo crev repo git rm -r ...` & `... commit` to delete the old ID. :)

**Reply by apoelstra** [reply]
In particular `cargo crev repo git rev-parse --show-toplevel` will show you where the actual git repo is, in case you want to do surgery.

**Reply by tcharding** [reply]
Awesome, I've cleaned up. Thanks
